### ISSUE TITLE
new lint: Unexpected double indirection

### ISSUE THREAD

**Issue by kornelski** [rust-lang-nursery/rust-clippy#3154]
In Rust it's relatively easy to accidentally end up with a `&&T` type when expecting to use `&T`.

The problem is that thanks to auto dereferencing `&&T` most of the time works just as well as `&T`, so the difference may go unnoticed for a while. But later when the code is changed, it may get weird type errors in unexpected places somewhere later in the code:

```rust
let s = "foo";
let s = &s[1..2]; // ok
let s = &s[1..2].trim(); // &&str suddenly
let l = s.len(); // but seems to work…

if s == "oo" // except, "can't compare `&str` with `str`"!?
```

```rust
for s in ["a", "b"].iter() { // gives &&str
    let s: String = s.to_owned(); // error: makes "owned" &str
}
```

So I'm hoping Clippy could find patterns which typically cause unexpected creation of doubly-indirect types, and warn about them before they become a hard-to-diagnose problem later.

**Comment by Manishearth** [comment]
I think we already have some lints which catch the _problematic_ cases here.

The issue with just blanket linting on this is that ephemeral double
references are actually really common and not problematic.

**Comment by kornelski** [comment]
There are no warnings for the aforementioned two tricky cases.

**Comment by kornelski** [comment]
There could be a narrower version of this lint, with an auto-fix, targeted for closure arguments in functions like `.filter(|arg|)` where the arg is a type like `&&str`. It could suggest using `.filter(|&arg|)` instead.
